hera_cal/abscal.py

```diff
@@ -3693,6 +3694,219 @@ def post_redcal_abscal_run(data_file, redcal_file, model_files, raw_auto_file=No
+                                inflate_model_by_redundancy=False, constrain_model_to_data_ants=False,
+                                clobber=False, tol=1e-6, max_iter=10,
+                                refant=None, ant_threshold=0.0, no_ampcal=False, no_phscal=False,
+                                dly_lincal=False,
+                                verbose=False):
+    """Driver function for model based calibration including i/o
+
+    Solve for gain parameters based on a foreground model.
+
+    Parameters
+    ----------
+    data_file: str
+        path to pyuvdata visibility file.
+        flags from this datafile are used for gains.
+    model_file: str
+        path to pyuvdata model file.
+    output_filename: str
+        path to output calibration file.
+    auto_file: str, optional
+        path to file containing autocorrelations and nsamples for inverse
+        variance weights. Default None -> use binary flag weights in calibration.
+    precalibration_gain_file: str, optional
+        Path to a gain file to apply to data before running calibration
+        default is None.
+    inflate_model_by_redundancy: bool, optional
+        expand model to match the redundant groups present in data file.
+        default is False.
+        Should be set to True if a redundant model is supplied or bad things
+        will happen!
+    constrain_model_to_data_ants: bool, optional
+        before inflating by redundancy, downselect array antennas in model to only
+        include antennas in the data. This avoids inflating model to full HERA array
+        which is memory inefficient for analyses using a small fraction of the array
+        but will break if the redundant baselines are keyed to antennas that are not
+        present in the data so only use this if you are confident that this is the case.
+        default is False.
+    clobber: bool, optional
+        overwrite outputs if True.
+    tol: float, optional
+        perform calibration loop until differences in gain solutions are at this level.
+    max_iter: int, optional
+        maximum number of iterations to perform
+        default is 10
+    refant: tuple, optional
+        referance antenna in form of (antnum, polstr).
+        Default is None -> automatically select a refant.
+    ant_threshold: float, optional
+        threshold of flags in frequency and time for a given antenna to completely
+        flag this antenna rom calibration.
+    no_ampcal: bool, optional
+        skip amplitude calibration (only calibrate phases)
+        default is False.
+    no_phscal: bool, optional
+        skip phase calibration (only calibrate amplitudes)
+        default is False.
+    dly_lincal: bool, optional
+        perform initial delay calibration before abscal
+        default is False.
+    verbose: bool, optional
+        lots of outputs.
+    """
+    hdd = io.HERAData(data_file)
```

Comment:
in this case the function `io.to_HERAData` is actually preferred b/c that allows the input `data_file` to be a str or a pre-loaded UVData object, which is nice for sandbox-mode analysis. could you make this change and update the doc string to reflect this?

Reply:
I'll change this. I also think that ideally `io.HERAData` ought to be able to take UVData objects ;)

Reply:
I've found that using strings in `to_HERAData` causes some very strange issues (see below). I'm going to stick with `io.HERAData(str)`

```
hera_cal/tests/test_abscal.py Fatal Python error: Illegal instruction

Current thread 0x000000010de52600 (most recent call first):
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pyuvdata/uvdata/aipy_extracts.py", line 295 in __init__
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pyuvdata/uvdata/miriad.py", line 837 in read_miriad
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pyuvdata/uvdata/uvdata.py", line 10623 in read_miriad
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pyuvdata/uvdata/uvdata.py", line 12164 in read
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pyuvdata/uvdata/uvdata.py", line 11826 in read
  File "/Users/aewallwi/Dropbox/Aaron/Science/HERA-Team/hera_cal/hera_cal/io.py", line 535 in read
  File "/Users/aewallwi/Dropbox/Aaron/Science/HERA-Team/hera_cal/hera_cal/abscal.py", line 3748 in run_model_based_calibration
  File "/Users/aewallwi/Dropbox/Aaron/Science/HERA-Team/hera_cal/hera_cal/tests/test_abscal.py", line 967 in test_run_model_based_calibration
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/python.py", line 183 in pytest_pyfunc_call
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/python.py", line 1641 in runtest
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/runner.py", line 162 in pytest_runtest_call
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/runner.py", line 255 in <lambda>
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/runner.py", line 311 in from_call
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/runner.py", line 254 in call_runtest_hook
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/runner.py", line 215 in call_and_report
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/runner.py", line 126 in runtestprotocol
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/runner.py", line 109 in pytest_runtest_protocol
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/main.py", line 348 in pytest_runtestloop
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/main.py", line 323 in _main
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/main.py", line 269 in wrap_session
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/main.py", line 316 in pytest_cmdline_main
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/config/__init__.py", line 162 in main
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/config/__init__.py", line 185 in console_main
  File "/Users/aewallwi/miniconda3/envs/hera/bin/pytest", line 8 in <module>
Illegal instruction: 4
```